### PR TITLE
[Gluten-4433][VL] Added cast conversion from int to timestamp

### PR DESCRIPTION
**Pull request by sharmaplkt**
Support for cast conversion from int to timestamp was missing.

Added functionality to convert int/string to timestamp.
Fixes: https://github.com/apache/incubator-gluten/issues/4433

### CODE REVIEW

**Comment by sharmaplkt** [comment]
@rui-mo @zhouyuan Can you review this PR.
Related PR for varchar to timestamp conversion.

File: velox/expression/CastExpr.cpp [hunk 1]
```diff
@@ -681,6 +705,8 @@ void CastExpr::applyPeeled(
     result = applyDecimal<int64_t>(rows, input, context, fromType, toType);
   } else if (toType->isLongDecimal()) {
     result = applyDecimal<int128_t>(rows, input, context, fromType, toType);
+  } else if (fromType->isBigint() && toType->isTimestamp()) {
+    result = castIntToTimestamp(rows, input, context);
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:709]: Maybe, we can support int8_t, int16_t, int32_t, int64_t with the same implementation like https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L639-L646.

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -247,6 +247,30 @@ VectorPtr CastExpr::castToDate(
...
+  auto* inputFlatVector = input.as<SimpleVector<int64_t>>();
+  const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+  const auto sessionTzName = queryConfig.sessionTimezone();
+  const auto* timeZone =
+      (queryConfig.adjustTimestampToTimezone() && !sessionTzName.empty())
+      ? date::locate_zone(sessionTzName)
+      : nullptr;
+  applyToSelectedNoThrowLocal(context, rows, castResult, [&](int row) {
+    auto timestamp = Timestamp::fromMillis(
+        inputFlatVector->valueAt(row) * kMillisInSecond);
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:267]: Would you align the implementation with longToTimestamp method in Spark? Overflow needs to be handled.
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L692-L693

File: velox/exec/tests/AggregationTest.cpp [hunk 3]
```diff
@@ -503,6 +503,24 @@ TEST_F(AggregationTest, missingFunctionOrSignature) {
...
+  auto intVector = makeFlatVector<int64_t>({2, 3, 4, 5});
+  auto timestampVector = makeFlatVector<Timestamp>(timeValues);
+  auto vectors = makeRowVector({"c1", "c2"}, {intVector, timestampVector});
+  createDuckDbTable({vectors});
+
+  auto op = PlanBuilder()
+                .values({vectors})
+                .project({"c1", "cast(c1 as timestamp)"})
+                .planNode();
+  assertQuery(op, "select c1,c2 from tmp");
```
Comment (rui-mo) [inline velox/exec/tests/AggregationTest.cpp:521]: Would you add some tests for overflow and with timezone? Please make sure the behaviors are aligned with Spark.

File: velox/exec/tests/AggregationTest.cpp [hunk 4]
```diff
@@ -503,6 +503,24 @@ TEST_F(AggregationTest, missingFunctionOrSignature) {
       "Supported signatures: (smallint,varchar) -> tinyint -> bigint.");
 }
 
+TEST_F(AggregationTest, castIntToTimestamp) {
```
Comment (rui-mo) [inline velox/exec/tests/AggregationTest.cpp:506]: Please move this test to functions/sparksql/tests/SparkCastExprTest.cpp.

File: velox/expression/CastExpr.cpp [hunk 5]
```diff
@@ -247,6 +247,30 @@ VectorPtr CastExpr::castToDate(
   }
 }
 
+VectorPtr CastExpr::castIntToTimestamp(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context) {
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:253]: Is the behavior the same between Spark and Presto? If not, we may need to separate the implementation.

File: velox/expression/CastExpr.h [hunk 6]
```diff
@@ -189,6 +189,11 @@ class CastExpr : public SpecialForm {
       exec::EvalCtx& context,
       const TypePtr& fromType);
 
+  VectorPtr castIntToTimestamp(
+      const SelectivityVector& rows,
+      const BaseVector& input,
+      exec::EvalCtx& context);
+
```
Comment (rui-mo) [inline velox/expression/CastExpr.h:196]: We need to update relating documentations.
For Presto, https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/presto/conversion.rst.
For Spark, https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/spark/conversion.rst.
If Presto and Spark behave the same, we can just update the documentation for Presto, and do not need to make a copy for Spark.

**Review by rui-mo — COMMENTED** [review]
Thanks!

File: velox/expression/CastExpr-inl.h [hunk 7]
```diff
@@ -285,6 +285,27 @@ void CastExpr::castTimestampToDate(
   });
 }
 
+template <typename IntType>
+void CastExpr::castIntegerToTimestamp(
+
```
Comment (rui-mo) [inline velox/expression/CastExpr-inl.h:290]: nit: remove this empty line.

File: velox/expression/CastExpr-inl.h [hunk 8]
```diff
@@ -285,6 +285,27 @@ void CastExpr::castTimestampToDate(
...
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    VectorPtr castResult,
+    const facebook::velox::date::time_zone* timeZone,
+    FlatVector<Timestamp>* resultFlatVector) {
+  applyToSelectedNoThrowLocal(context, rows, castResult, [&](int row) {
+    auto timestamp = Timestamp::fromMillis(
+        static_cast<int64_t>(input.as<SimpleVector<IntType>>()->valueAt(row)) *
```
Comment (rui-mo) [inline velox/expression/CastExpr-inl.h:299]: I wonder if the conversion to int64_t is implicit and we may not need static_cast.

File: velox/expression/CastExpr-inl.h [hunk 9]
```diff
@@ -285,6 +285,27 @@ void CastExpr::castTimestampToDate(
...
+
+        kMillisInSecond);
```
Comment (rui-mo) [inline velox/expression/CastExpr-inl.h:301]: In find in Spark, `SECONDS.toMicros(t)` is implemented as below where `x` will check the long value against a maximum value. Do we need similar implementation in Velox?
```
   public long toMicros(long var1) {
            return x(var1, 1000000L, 9223372036854L);
        }

    static long x(long var0, long var2, long var4) {
        if (var0 > var4) {
            return Long.MAX_VALUE;
        } else {
            return var0 < -var4 ? Long.MIN_VALUE : var0 * var2;
        }
    }
```

File: velox/expression/CastExpr.cpp [hunk 10]
```diff
@@ -248,6 +248,46 @@ VectorPtr CastExpr::castToDate(
...
+          rows, input, context, castResult, timeZone, resultFlatVector);
+      break;
+    case TypeKind::BIGINT:
+      castIntegerToTimestamp<int64_t>(
+          rows, input, context, castResult, timeZone, resultFlatVector);
+      break;
+    default:
+      VELOX_UNREACHABLE(
+          "Cast from {} to Timestamp is not supported", fromType->kind());
+      break;
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:286]: nit: break is not needed.

File: velox/expression/CastExpr.cpp [hunk 11]
```diff
@@ -248,6 +248,46 @@ VectorPtr CastExpr::castToDate(
...
+  VectorPtr castResult;
+  context.ensureWritable(rows, TIMESTAMP(), castResult);
+  (*castResult).clearNulls(rows);
+  auto* resultFlatVector = castResult->as<FlatVector<Timestamp>>();
+  const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+  const auto sessionTzName = queryConfig.sessionTimezone();
+  const auto* timeZone =
+      (queryConfig.adjustTimestampToTimezone() && !sessionTzName.empty())
+      ? date::locate_zone(sessionTzName)
+      : nullptr;
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:265]: It seems we can combine above code into castIntegerToTimestamp, and dispatch in applyPeeled once. Do you think that would make the implementation more concise?

File: velox/expression/tests/CastExprTest.cpp [hunk 12]
```diff
@@ -1186,32 +1177,6 @@ TEST_F(CastExprTest, mapCast) {
...
-        inputMap,
-        makeMapVector<Timestamp, int64_t>(
-            kVectorSize,
-            sizeAt,
-            [](auto /*row*/) { return Timestamp(); },
-            valueAt,
-            [](auto row) { return row % 3 == 0 || row % 5 != 0; }),
-        true);
-  }
-
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:1214]: I wonder why we need to remove these tests. Can we keep them?

**Review by rui-mo — COMMENTED** [review]
@sharmaplkt Thanks for your continuous work. Just added several comments.

**Pull request closed without merging**